src/src/PluginStructs/P014_data_struct.cpp

```diff
@@ -1,144 +1,218 @@
-      }
-      break;
-    }
-    case SI7021_state::Values_read:
-    {
-      // This must be done in a separate call to
-      // make sure we only start reading when the plugin wants us to perform a reading.
-      // Change state of sensor for non blocking reading
-      state = SI7021_state::Initialized;
-      break;
-    }
-  }
-  return success;
+//The main state machine
+//Only perform the measurements with big interval to prevent the sensor from warming up.
+//returns true when the state changes and false otherwise
+bool P014_data_struct::update(uint8_t i2caddr, uint8_t resolution, uint8_t filter_power) {
+  //const unsigned long current_time = millis();
+  //int8_t ret=0;
+  
+  switch(state){
+    case P014_state::Uninitialized:
+        //make sure we do not try to initialize too often
+        //someone could enable the plugin but does not have the hardware connected
+        if (!timeOutReached(last_measurement_time + errCount*SI70xx_INIT_DELAY)) {
+          return false;
+        }
+
+        //we have to stop trying after a while
+        if (errCount>P014_MAX_RETRY){
+            state = P014_state::Error;
+            return true;
+        }
+
+        if (I2C_wakeup(i2caddr)!=0){
+          String log = F("SI70xx : Not available at address: ");
+          log += String(i2caddr, HEX);
+          addLog(LOG_LEVEL_ERROR, log);
+          errCount++;
+          last_measurement_time = millis();
+          return false;
+        }
+
+        if (init(i2caddr, resolution)){
+          state                 = P014_state::Initialized;
+          last_measurement_time = millis();
+          errCount              = 0;
+          return true;
+        }
+
+        last_measurement_time = millis();
+        errCount++;
+        return false;
+    break;
+
+    case P014_state::Initialized:
+      if (chip_id == CHIP_ID_SI7013){
+        //we need to wait for the read of ADC
+        if (!timeOutReached(last_measurement_time + SI70xx_MEASUREMENT_DELAY)) {
+          return false;
+        }
+
+        if ( !readADC(i2caddr,filter_power)){
+          return false;
+        }
+        
+        adc = adc << filter_power; //this is the first measurement
+      } 
+      last_measurement_time = millis();
+      state = P014_state::Ready;
+      return true;
+    break;
+
+    case P014_state::Ready:
+        if (!I2C_write8(i2caddr,SI70xx_CMD_MEASURE_HUM)) { //measure humidity and temperature at the same time
+          addLog(LOG_LEVEL_ERROR,F("SI70xx: startConv Failed!"));
+          return false;
+        }
+      
+        last_measurement_time = millis();
+        state = P014_state::Wait_for_humidity_samples;
+    break;
+
+    case P014_state::Wait_for_humidity_samples:
+        //make sure we wait for the measurement to complete
+        if (!timeOutReached(last_measurement_time + SI70xx_MEASUREMENT_DELAY)) {
+          return false;
+        }
+
+        if (!readHumidity(i2caddr, resolution)) {
```

Comment:
`readHumidity` does also return false when there are less than 3 bytes to read.
So what happens if there are only 1 or 2 bytes read?
The reading isn't restarted as the state is not changed.
But the next call to read the humidity might read 3 bytes of which some are not related to the humidity thus the checksum will fail.
Then also the state is not changed.
I think the plugin might never get out of this loop.

Reply:
maybe my understanding is wrong but I2C has an internal buffer that is reused.
On the next cycle will read the missing byte and keep the previous bytes from the buffer.
I have not tested that assumption however if that is wrong then yes I need to go back one step at least or even reset the device as most likely the device is going to be  in an inconsistent state.
This never happened to me even with 1 sec interval. So not sure if I need to change anything.

Reply:
What will be in the buffer if another I2C device is being read while you are 'away'?

Reply:
note taken. I am moving it back to request a read.